### DISCUSSION TITLE
Version 6.4.1 breaks my code

### DISCUSSION THREAD

**Discussion by mmick66** [mikro-orm/mikro-orm#6285] [General]
I just ran a routing NPM update and got version `6.4.1`. I immediately started seeing errors popping everywhere. I will give one brief example here.

```typescipt
const populateForOptions = (options: UserFindOptions): { populate: Array<string> } | undefined => {
  const populate = [];

  /* always include roles */
  populate.push(nameof<UserEntity>((u) => u.roles));

  if (options.profile) {
    populate.push(
      nameof<UserEntity>((u) => u.profile)
    );
  }

  return { populate };
};

async findById(id: string, options?: UserFindOptions): Promise<UserEntity> {
  return this.userRepo.findOne(
    { id },
    {
      ...populateForOptions(options)
    }
  );
}
```
The error is 
```
TS2345: Argument of type { populate: Array<string>; } is not assignable to parameter of type
FindOneOptions<UserEntity, string, '*', never>
Types of property populate are incompatible.
Type string[] is not assignable to type Populate<UserEntity, string>
Type string[] is not assignable to type readonly never[]
Type string is not assignable to type never
```

Please advise.

**Comment by B4nan** [comment]
You will need to provide a complete reproduction, I can't help you without that. My guess is you upgraded TS too, I am not aware of any changes in public methods on type level.

**Reply by mmick66** [reply]
Thank you. I think it's not part of your library.

**Comment by mmick66** [comment]
Ok, so here's the thing. I don't know what on earth happened! Typescript got updated without my knowledge and all hell broke loose. Errors started popping up everywhere. This does not concern you, of course. What *does* concern you is that in your `Populate` parameter I now get errors 
```bash
TS2322: Type 'string' is not assignable to type 'never'.
```
The reason is that instead of passing string literals like `populate: ['profile']` I have a function to generate them. Look at this simple code snipped
```
const getRelations = () => 'profile';
return this.userRepo.findOne(
  { username: regex },
  { populate: getRelations() }
);
```
This throws an error now
```
TS2322: Type string is not assignable to type Populate<UserEntity, never>
IDatabaseDriver. d. ts(72, 5): The expected type comes from property populate which is declared here on type FindOneOptions<UserEntity, never, '*', never>
```
The `userRepo` is just an instance of your `EntityRepository` (injected in NestJS)

Something happened with type inference in the latest typescript update that does not play well with your types. I can change my code, but I believe that the snippet above should just work, as it did until Friday!

Let me know if you have any thoughts.

**Reply by B4nan** [reply]
I am actually quite confused this worked for you before, I don't think it should. You can't pass `string` to the populate hint (and a helper like `const getRelations = () => 'profile'` will always return a string, that didn't change with the latest TS update), it needs to be a literal string union, so `profile`. Hard to guess without a repro, but from what you provided here, I'd guess it will be about you changing your code, the input types need to be correct.

I think that they just improved how the inference works, and you might have been getting `any` somewhere before, now you get `string`. If that is the case, the TS update just uncovered that you were relying on `any` instead of strictness.

**Reply by B4nan** [reply]
If you share an actual example of your helper function, maybe I can help you make it properly typed so you don't end up with just `string`, at least I hope so.

For that static example, `as const` would fix it:

```ts
const getRelations = () => 'profile' as const;
```

**Reply by mmick66** [reply]
Thank you for taking the time. I am using this function, whichI copied verbatim, to help me with custom props.

https://github.com/IRCraziestTaxi/ts-simple-nameof/blob/master/src/nameof.ts

Is it just about returning `as const` ?

**Reply by B4nan** [reply]
Huh, so you are trying to extract a type name out of a code string or something like that? I don't think that can work, all solutions to that I saw were hacks, so I wouldn't be surprised things like this break with TS upgrades.

I think you're best bet is to type it as `any` instead of `string`, not much you can do there.

Why are you trying to do this?

**Reply by mmick66** [reply]
Well, so that I can get some "type safety." If I have a type `Students` and I want it's `.teacher` property, I can just use the function above and get a `nameof<Student>((s) => s.teacher))`, It's just a utility function really.

Thank you for your help once again.

**Reply by B4nan** [reply]
The `populate` option is already typesafe, there is no point in doing this if you ask me.

**Reply by mmick66** [reply]
Yes! We just realised that we where trying to add safety where we thought (wrongly) there was none.

**Reply by mmick66** [reply]
By the way, since we are at it, what is the best way to write a function that returns something that you could pass into the `{ populate: .. }` property?

**Reply by B4nan** [reply]
Depends on what you want from it, but if it should be dynamic, it needs to use generics for the literal string value, maybe `const` type arguments could work here too.
